### PR TITLE
[CIP-50] optional replay protection

### PR DESCRIPTION
**Pull request by gastonponti**
Depends on #1695

### Description

Implementation of the CIP 50, that makes the replay protection optional

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

No. Part of the Expresso hard fork

### CODE REVIEW

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 32ddca7a638f53ab6721c643cf0782c13f6de7e0
<details><summary>coverage:  44.6% of statements across all listed packages</summary><br>
<pre>
coverage:  54.1% of statements in consensus/istanbul
coverage:  44.7% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  43.3% of statements in consensus/istanbul/backend/internal/db
coverage:  24.1% of statements in consensus/istanbul/backend/internal/enodes
coverage:  22.4% of statements in consensus/istanbul/backend/internal/replica
coverage:  63.9% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
</pre>
CommentID: 09775b3d7b
</details>

**Review by trianglesphere — APPROVED** [review]
Blockchain side LGTM. Will need to update CIP-35 e2e test.

File: core/tx_pool_test.go [hunk 1]
```diff
@@ -876,42 +876,72 @@ func TestTransactionGapFilling(t *testing.T) {
...
 	pool, key := setupTxPool()
+	// Create a test account and fund it
 	defer pool.Stop()
 
 	account := crypto.PubkeyToAddress(key.PublicKey)
 	pool.currentState.AddBalance(account, big.NewInt(1000000))
 
+	// flag it as before donut
+	pool.donut = false
+	pool.eHardfork = false
```
Comment (trianglesphere) [inline core/tx_pool_test.go:891]:
```suggestion
	pool.espresso = false
```

File: core/tx_pool_test.go [hunk 2]
```diff
@@ -876,42 +876,72 @@ func TestTransactionGapFilling(t *testing.T) {
...
-	if queued != 4 {
-		t.Fatalf("queued transactions mismatched: have %d, want %d", queued, 4)
+	if queued != 3 {
+		t.Fatalf("after donut, queued transactions mismatched: have %d, want %d", queued, 3)
 	}
 
-	pool.handleDonutActivation()
+	// In E fork
+	// flag it as E hard fork
+	pool.eHardfork = true
```
Comment (trianglesphere) [inline core/tx_pool_test.go:930]:
```suggestion
	pool.espresso = true
```

File: core/tx_pool_test.go [hunk 3]
```diff
@@ -876,42 +876,72 @@ func TestTransactionGapFilling(t *testing.T) {
...
+
+		transaction(13, 100000, key),
+		protectedTransaction(14, 100000, key),
+	})
 
 	pending, queued = pool.Stats()
-	if pending != 1 {
-		t.Fatalf("pending transactions mismatched: have %d, want %d", pending, 1)
+	if pending != 5 {
+		t.Fatalf("after expresso, pending transactions mismatched: have %d, want %d", pending, 5)
```
Comment (trianglesphere) [inline core/tx_pool_test.go:941]:
```suggestion
		t.Fatalf("after espresso, pending transactions mismatched: have %d, want %d", pending, 5)
```

File: core/tx_pool_test.go [hunk 4]
```diff
@@ -876,42 +876,72 @@ func TestTransactionGapFilling(t *testing.T) {
...
 	}
-	if queued != 2 {
-		t.Fatalf("queued transactions mismatched: have %d, want %d", queued, 2)
+	if queued != 5 {
+		t.Fatalf("after expresso, queued transactions mismatched: have %d, want %d", queued, 5)
```
Comment (trianglesphere) [inline core/tx_pool_test.go:944]:
```suggestion
		t.Fatalf("after espresso, queued transactions mismatched: have %d, want %d", queued, 5)
```

**Review by trianglesphere — COMMENTED** [review]
Update around espresso name in tests

**Pull request merged**